src/realm/sync/instruction_replication.cpp

```diff
@@ -424,12 +424,31 @@ void SyncReplication::list_set(const CollectionBase& list, size_t ndx, Mixed val
     }
 
     if (select_collection(list)) {
-        Instruction::Update instr;
-        populate_path_instr(instr, list, uint32_t(ndx));
-        REALM_ASSERT(instr.is_array_update());
-        instr.value = as_payload(list, value);
-        instr.prior_size = uint32_t(list.size());
-        emit(instr);
+        // If this is an embedded object then we need to emit and erase/insert instruction so that the old
+        // object gets cleared, otherwise you'll only see the Update ObjectValue instruction, which is idempotent,
+        // and that will lead to corrupted prior size for array operations inside the embedded object during
+        // changeset application.
+        if (value.is_type(type_Link, type_TypedLink) && list.get_target_table()->is_embedded()) {
```

Comment:
If the type is a list of mixed, then `list.get_target_table()` will return null. In that case, the mixed value will be a typed link and we should get the target table from `value.get_link().get_table_key()`. Could you add a test for this?

Reply:
Done. I now handle type_Link and type_TypedLink separately.